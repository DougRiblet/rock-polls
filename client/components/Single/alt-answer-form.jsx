// @flow

import React from 'react';

type Props = {
  handleAlt: (string, string) => mixed,
  pollid: string,
};

type State = {
  answer: string,
};

export default class Login extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      answer: '',
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleSubmit: Function;
  handleChange: Function;

  // eslint-disable-next-line no-undef
  handleChange(event: SyntheticInputEvent<*>) {
    this.setState({ answer: event.target.value });
  }

  // eslint-disable-next-line no-undef
  handleSubmit(event: SyntheticInputEvent<*>) {
    event.preventDefault();
    this.props.handleAlt(this.props.pollid, this.state.answer);
    this.setState({ answer: '' });
  }

  render() {
    return (
      <div id='alt-answer'>
        <form onSubmit={this.handleSubmit}>
          <label>
            <input
              id='alt-input'
              type='text'
              value={this.state.answer}
              maxLength='99'
              size='80'
              onChange={this.handleChange}
            />
          </label>
          <input type='submit' value='submit' />
        </form>
      </div>
    );
  }
}
