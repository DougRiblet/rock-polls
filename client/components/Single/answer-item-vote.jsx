// @flow

import React from 'react';

type Props = {
  handleVote: (string) => mixed,
  aId: string,
  aInfo: { answer: string, count: number},
};

const AnswerItemVote = ({ aId, aInfo, handleVote }: Props) => {
  const handleClick = () => handleVote(aId);
  return (
    <li>
      <button className='answer-item-vote' onClick={handleClick} >
        { aInfo.answer }
      </button>
    </li>
  );
};

export default AnswerItemVote;
