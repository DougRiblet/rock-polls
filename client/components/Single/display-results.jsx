// @flow

import React from 'react';

type Props = {
  allPolls: Object,
  allAnswers: Object,
  pollid: string,
};

const DisplayResults = ({ allPolls, allAnswers, pollid }: Props) => {
  const getDisplay = () => {
    const { answers } = allPolls[pollid];
    if (answers) {
      return answers.map((aId) => {
        if (allAnswers[aId]) {
          const aObj = allAnswers[aId];
          aObj.id = aId;
          return aObj;
        }
        return null;
      })
        .filter(y => y)
        .sort((a, b) => a.count < b.count)
        .map(a => <li key={a.id}>{a.answer} - {a.count}</li>);
    }
    return '';
  };
  return (
    <ul>
      { getDisplay() }
    </ul>
  );
};

export default DisplayResults;
