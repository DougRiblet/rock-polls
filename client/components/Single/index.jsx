// import { connect } from 'react-redux';
// import { bindActionCreators } from 'redux';
// import { grabAllPolls } from '../../actions/actions';
// import singlePresent from './single-presentational';

// const mapStateToProps = state => ({
//   allPolls: state.allpolls,
// });

// const mapDispatchToProps = dispatch => bindActionCreators({ grabAllPolls }, dispatch);

// export default connect(mapStateToProps, mapDispatchToProps)(singlePresent);

import React from 'react';

const Single = ({ match }) => (
  <div>
    <h2>## SINGLE POLL: { match.params.pollid }</h2>
  </div>
);

export default Single;
