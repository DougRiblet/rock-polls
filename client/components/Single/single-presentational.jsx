// @flow

import React from 'react';
import AnswerItemVote from './answer-item-vote';

type Props = {
  castVote: (answerId) => mixed,
  grabSinglePoll: (pollId) => mixed,
  authenticated: boolean,
  user_id: string,
  allPolls: Object,
  allAnswers: Object,
};

type State = {
  hasVoted: boolean,
};

export default class Single extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      hasVoted: false,
    };
  this.displayQuestion = this.displayQuestion.bind(this);
  this.displayAnswers = this.displayAnswers.bind(this);
  this.handleVote = this.handleVote.bind(this);
  }

  componentDidMount() {
    this.props.grabSinglePoll(this.props.match.params.pollid);
  }

  displayQuestion(pollid) {
    return this.props.allPolls[pollid].question;
  }

  displayAnswers(pollid) {
    const answers = this.props.allPolls[pollid].answers;
    if (answers) {
      return answers.map(aId => {
        const aInfo = this.props.allAnswers[aId];
        return <AnswerItemVote
            key={aId}
            aId={aId}
            aInfo={aInfo}
            handleVote={this.handleVote}
          />
        // return <li
        //   key={answerId}
        //   name={answerId}
        //   onClick={() => this.handleVote(answerId)}
        // >
        //   { a.answer } - { a.count }
        // </li>;
      });
    }
    return <li></li>
  }

  // eslint-disable-next-line no-undef
  handleVote(answerid) {
    this.props.castVote(answerid);
    this.setState({ hasVoted: true });
  }

  render() {
    const pollid = this.props.match.params.pollid;
    return (
      <div id='single'>
        <h2>
          { this.displayQuestion(pollid) }
        </h2>
        <ul>
          { this.displayAnswers(pollid) }
        </ul>

      </div>
    );
  }
}

