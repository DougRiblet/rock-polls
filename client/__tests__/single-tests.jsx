import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Single from '../components/Single/single-presentational';

jest.mock('react-router-dom', () => ({ Link: 'Link' }));

describe('Single', () => {
  test('Snapshot', () => {
    const tree = renderer.create(<Single
      castVote={() => jest.fn()}
      grabSinglePoll={() => jest.fn()}
      addAltAnswer={() => jest.fn()}
      allPolls={{}}
      allAnswers={{}}
      authenticated={false}
      user_id=''
      match={{ params: { pollid: 'abcdef' } }}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Single
      castVote={() => jest.fn()}
      grabSinglePoll={() => jest.fn()}
      addAltAnswer={() => jest.fn()}
      allPolls={{}}
      allAnswers={{}}
      authenticated={false}
      user_id=''
      match={{ params: { pollid: 'abcdef' } }}
    />, div);
  });
});
