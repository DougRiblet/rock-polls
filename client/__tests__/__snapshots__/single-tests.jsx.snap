// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Snapshot 1`] = `
<div
  className="single"
>
  <div
    className="backlink"
  >
    <Link
      to="/"
    >
      back to polls list
    </Link>
  </div>
  <div
    className="single-body"
  >
    <h1>
      
    </h1>
    <div>
      <ul>
        
      </ul>
      <div>
        <p
          className="alt-suggest"
        >
          Click your choice to vote. Logged-in users have a write-in option.
        </p>
      </div>
    </div>
  </div>
</div>
`;
